subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java

```diff
@@ -154,20 +161,30 @@ private boolean isClass(TestIdentifier test) {
     }
 
     private String className(TestIdentifier testIdentifier) {
+        return classNameAndDisplayName(testIdentifier).className;
+    }
+
+    private ClassNameAndDisplayName classNameAndDisplayName(TestIdentifier testIdentifier) {
         // For tests in default method of interface,
         // we might not be able to get the implementation class directly.
         // In this case, we need to retrieve test plan to get the real implementation class.
         if (isClass(testIdentifier)) {
-            return ClassSource.class.cast(testIdentifier.getSource().get()).getClassName();
+            return extractClassNameAndDisplayName(testIdentifier);
```

Comment:
⭕️ This can probably be refactored so we don't need an additional data object. I.e. first find the identifier you care about, then extract its class name and display name. Not a big deal though.

Reply:
Yes, I don't like the dedicated data class either. Changed as you say.